### PR TITLE
Implement a qkv formulation using additive attention

### PR DESCRIPTION
**Pull request by mtanghu**
The original Fastformer structure may not quite be right for Causal Language Modeling/flexible decoder based modeling for all purposes. This would be implicitly evidenced by having to introduce the residual connection and layer norm to get the original Fastformer to converge.

This pull request simply introduces a query, key, value formulation of using Additive Attention that is much simpler to show/explain and also gives the ability for tokens to "choose whether to pay attention to the cumsum-ed attention vector".

If this project chooses to go this route, it would be recommendable to still leave the original version in a folder in say the src folder, and abandon the name "fastformer" as the structure/formulations are different enough to use a different name where the only similarity would just be the concept of "Additive Attention".

Would also be good to explore getting rid of how the attention weights are generated with a learned *vector* this is likely a pretty strong bottleneck as to the representation capacity of this model.

### CODE REVIEW

**Comment by mtanghu** [comment]
This series of commits gets rid of the "learned vector" and replaces it with something much closer to full attention (with likewise much more representational complexity). Which is also much cleaner

In this way we move away from Additive Attention completely (and Fastformer for that matter). Likewise this project will be rebranding to be called "Linear Explainable Attention with Parallelization" (or LEAP) which uses "Focused Attention"

- [x] Put old fastformerLM code in a contained folder in src (so people can still use/try it if they want)
- [x] Fix the structure in the archived repo to follow GPT2 (update testing and README too)
- [x] Edit old README to mention that a newer approach has been taken and remove the benefits and development section. Make sure the new focus is geared toward users who just want to try fastformer for language modeling 
- [x] Remove embedding layer norm, don't scale last layer by srqt(d_model), add a layer norm at the end before projecting logits (this all to fit attention is all you need and likewise gpt2). Don't worry about multiplying the embedding by sqrt(d_model) the huggingface implementation doesn't have this even though the paper does call for it.
- [x] Redo fastformer using strong scaling technique and getting rid of the residual connection

**Comment by mtanghu** [comment]
- [ ] Write a new README with new math and the benefit/development sections from old README, also mentioning the legacy fastformerLM that is still in the repo. Make sure to note that for now the focus is on masked attention for decoders.
- [ ] Try replacing gating with head-wise cosine similarity? -- I think this may have better backwards gradient flow and increase explainability (over simple gating). I suppose then you would have 2 "key" vectors -- think about renaming these so the math makes sense
- [ ] Make a short section on STRONG scaling for both readmes
- [ ] Make a dedicated LEAP attention block that people can use, document this in new README when you have time (I suppose for now it will be decoder only, but you can easily make it bidirectional by reversing the token order and running it through again)
- [ ] Put a section about development/contributing making sure to tell people about pip install -e .

**Comment by mtanghu** [comment]
Still some work to be done to finish the renovations, the work done here is just the groundwork see #8 for more information on the QKV renovations
